containers/src/Data/IntMap/Internal.hs

```diff
@@ -353,26 +356,45 @@ intFromNat = fromIntegral
 
 -- See Note: Order of constructors
 data IntMap a = Bin {-# UNPACK #-} !Prefix
-                    {-# UNPACK #-} !Mask
                     !(IntMap a)
                     !(IntMap a)
--- Fields:
---   prefix: The most significant bits shared by all keys in this Bin.
---   mask: The switching bit to determine if a key should follow the left
---         or right subtree of a 'Bin'.
--- Invariant: Nil is never found as a child of Bin.
--- Invariant: The Mask is a power of 2. It is the largest bit position at which
---            two keys of the map differ.
--- Invariant: Prefix is the common high-order bits that all elements share to
---            the left of the Mask bit.
--- Invariant: In (Bin prefix mask left right), left consists of the elements that
---            don't have the mask bit set; right is all the elements that do.
               | Tip {-# UNPACK #-} !Key a
               | Nil
 
-type Prefix = Int
-type Mask   = Int
+-- IntMap structure and invariants:
+--
+-- * Nil is never found as a child of Bin.
+--
+-- * The Prefix of a Bin indicates the common high-order bits that all keys in
+--   the Bin share.
+--
+-- * The least significant set bit of the Int value of a Prefix is called the
+--   mask bit.
+--
+-- * All the bits to the left of the mask bit are called the shared prefix. All
+--   keys stored in the Bin begin with the shared prefix.
+--
+-- * All keys in the left child of the Bin have the mask bit unset, and all keys
+--   in the right child have the mask bit set. It follows that
+--
+--   1. The Int value of the Prefix of a Bin is the smallest key that can be
+--      present in the right child of the Bin.
+--
+--   2. All keys in the right child of a Bin are greater than keys in the
+--      left child, with one exceptional situation. If the Bin separates
+--      negative and non-negative keys, the mask bit is the sign bit and the
+--      left child stores the non-negative keys while the right child stores the
+--      negative keys.
+--
+-- * All bits to the right of the mask bit are set to 0 in a Prefix.
+--
 
+-- See Note [Okasaki-Gill] for how the implementation here relates to the one in
+-- Okasaki and Gill's paper.
+
+-- | A @Prefix@ represents some prefix of high-order bits of an @Int@.
+newtype Prefix = Prefix { unPrefix :: Int }
+  deriving (Eq, Lift)
```

Comment:
Yeah, that's altogether fine.

Reply:
Deriving `Lift` instances is almost always the right thing. The only exceptions I can think of are:

1. Data structures with really weird things like `IORef`s, `TVar`s, pointer cycles, or infinite/very large lazily generated structures. These require special treatment to lift, when they can be lifted at all.
2. Redundant data structures where a particular canonical form is more desirable. This is the case for `Data.Sequence.Seq`, where the `Lift` instance puts its argument into perfectly balanced left-heavy form. I don't remember what we do for `Map` and `Set`, but we might also have special logic for those.

Note that you don't really have to worry too much about `Lift` breaking abstraction barriers, because Template Haskell already smashes those to smithereens.

Reply:
Hmmm.... Actually..... Now you're reminding me of another issue, which is code size of lifted structures. Sometimes it's possible to reduce that (e.g., by consolidating coercions) using a custom instance. In principle, we could write a `Lift` instance for `IntMap` that uses a "smart constructor" instead of `Bin`, to take an `Int` instead of a `Prefix`. We'd have to see if that actually helped in the end or not.

Reply:
By code size, I actually meant Core term size, which can affect compilation time and such. It doesn't directly affect the compiled code size.

Reply:
tl;dr: the newtype you've introduced will almost certainly make some things slower at compile time (type checking and optimizing splices generated by `lift`). But the only workarounds I can think of will either slow down runtime code or use `unsafeCoerce` in an extremely shady way that I'm guessing would affect profiling. So don't worry about it.

Reply:
Got it 👍